docker/bookkeeper/entrypoint.sh

```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
-  # The container already contains data in BK directories. This is probably because
-  # the container has been restarted; or, if running on Kubernetes, it has probably been
-  # updated or evacuated without losing its persistent volumes.
-  echo "data available in bookkeeper directories; not formatting the bookie"
-else
-  # The container does not contain any BK data, it is probably a new
-  # bookie. We will format any pre-existent data and metadata before starting
-  # the bookie to avoid potential conflicts.
-  echo "format bookie data and metadata"
-  /opt/bookkeeper/bin/bookkeeper shell bookieformat -nonInteractive -force -deleteCookie
-fi
-
-echo "start bookie"
-/opt/bookkeeper/scripts/entrypoint.sh bookie
+# Create directories for multiple ledgers and journals if specified.
+create_bookie_dirs() {
+  IFS=',' read -ra directories <<< $1
+  for i in "${directories[@]}"
+  do
+      mkdir -p $i
+      if [ "$(id -u)" = '0' ]; then
+          chown -R "${BK_USER}:${BK_USER}" $i
+      fi
+  done
+}
+
+wait_for_zookeeper() {
+    echo "Waiting for zookeeper"
+    until zk-shell --run-once "ls /" ${BK_zkServers}; do sleep 5; done
+}
+
+
+create_zk_root() {
+    if [ "x${BK_CLUSTER_ROOT_PATH}" != "x" ]; then
+        echo "Creating the zk root dir '${BK_CLUSTER_ROOT_PATH}' at '${BK_zkServers}'"
+        zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH} '' false false true" ${BK_zkServers}
+    fi
+}
+
+configure_bk() {
+    # We need to update the metadata endpoint and Bookie ID before attempting to delete the cookie
+    sed -i "s|.*metadataServiceUri=.*\$|metadataServiceUri=${BK_metadataServiceUri}|" /opt/bookkeeper/conf/bk_server.conf
+    if [ ! -z "$BK_useHostNameAsBookieID" ]; then
+      sed -i "s|.*useHostNameAsBookieID=.*\$|useHostNameAsBookieID=${BK_useHostNameAsBookieID}|" ${BK_HOME}/conf/bk_server.conf
+    fi
+}
+
+fix_bk_ipv6_check() {
+
+   # `${SCRIPTS_DIR}/common.sh` - a shell script in Bookeeper, runs this command: `/sbin/sysctl -n net.ipv6.bindv6only`.
+   # This command can return an error exit code `255` and error message:
+   #    `sysctl: cannot stat /proc/sys/net/ipv6/bindv6only: No such file or directory`
+   #
+   # If the `common.sh` file is sourced in a script that sets "exit on error" as true (set -e), the script shall return
+   # an error code `255`. Since, the `bookkeeper/bin/bookeeper` shell script sources the `common.sh` file, and it sets
+   # "exit on error" as true, we encounter this error when it is invoked. To avoid that error, here we modify the
+   # `common.sh` file to execute `/sbin/sysctl -n net.ipv6.bindv6only` only if file `/proc/sys/net/ipv6/bindv6only`
+   # is available.
+
+   # Replace the 22nd line with "if [ -f /sbin/sysctl ] && [ -f /proc/sys/net/ipv6/bindv6only ]; then".
+   sed -i "22s|.*|if [ -f /sbin/sysctl ] \&\& [ -f /proc/sys/net/ipv6/bindv6only ]; then|" ${BK_HOME}/bin/common.sh
```

Comment:
it looks like we can propose a change upstream to Apache BookKeeper rather than making patches like this. Are you considering doing it?

Reply:
Yes, there is an open PR for `apache/bookkeeper`: https://github.com/apache/bookkeeper/pull/2219.